File: src/test/kotlin/com/example/mortgage/service/CustomerServiceImplTest.java
package com.example.mortgage.service;

import com.example.mortgage.model.Customer;
import com.example.mortgage.repository.CustomerRepository;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.junit4.SpringRunner;

import java.util.List;
import java.util.Optional;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

@RunWith(SpringRunner.class)
@SpringBootTest
public class CustomerServiceImplTest {

    @Autowired
    private CustomerRepository repository;

    @Autowired
    private CustomerService service;

    private final long id = 4L;


    @Test
    public void testFindById() {

        Optional<Customer> entity = service.getCustomerById(id);

        assertTrue(entity.isPresent());

        Customer customer = entity.get();

        assertEquals(id, customer.getCustomerId());
        assertEquals("Almost", customer.getFirstName());
        assertEquals("There", customer.getLastName());
        assertNull(customer.getPhone());
        assertNull(customer.getEmail());
    }

    @Test
    public void testFindAll() {

        List<Customer> customers = service.getCustomers();

        assertEquals(4, customers.size());

        for (Customer customer : customers) {
            System.out.println("customer = " + customer);
        }
    }

    @Test
    public void testUpdateCustomer() {

        final Customer customer = new Customer.Builder()
            .withCustomerId(id)
            .withFirstName("Yogi")
            .withLastName("Bear")
            .withPhone("8005550001")
            .withEmail("main.bear@jellystone.gov")
            .build();

        final Customer serviceResult = service.updateCustomer(customer);
        assertEquals(customer, serviceResult);

        final Optional<Customer> entity = repository.findById(id);
        assertTrue(entity.isPresent());
        assertEquals(customer, entity.get());
    }

    @Test
    public void testUpdateCustomer_invalidId() {

        final long invalidId = 9999;
        final Customer customer = new Customer.Builder()
            .withCustomerId(invalidId)
            .withFirstName("Yogi")
            .withLastName("Bear")
            .withPhone("8005550001")
            .withEmail("main.bear@jellystone.gov")
            .build();

        final Customer serviceResult = service.updateCustomer(customer);
        assertNull(serviceResult);

        final Optional<Customer> entity = repository.findById(invalidId);
        assertFalse(entity.isPresent());
    }

    @Test
    public void testAddAndDeleteCustomer() {

        final Customer customer =
            new Customer.Builder().withFirstName("Teen").withLastName("Titans").withPhone("8005557777").build();

        final long expectedCount = repository.count() + 1;
        final Customer entity = service.addCustomer(customer);

        assertEquals(expectedCount, repository.count());

        customer.setCustomerId(entity.getCustomerId());
        assertEquals(customer, entity);

        service.deleteCustomer(entity.getCustomerId());
        assertEquals(expectedCount - 1, repository.count());
        assertFalse(service.getCustomerById(entity.getCustomerId()).isPresent());
    }

}

File: src/test/kotlin/com/example/mortgage/service/CustomerServiceImplTest.kt
package com.example.mortgage.service

import com.example.mortgage.digest.CustomerDigest
import com.example.mortgage.model.Customer
import com.example.mortgage.repository.CustomerRepository
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.junit4.SpringRunner

@RunWith(SpringRunner::class)
@SpringBootTest
class CustomerServiceImplTest {
    @Autowired
    lateinit var repository: CustomerRepository

    @Autowired
    lateinit var service: CustomerService

    private val id = 4L

    @Before
    fun initialize() {
        val customers = repository.findAll()
        for (customer in customers) {
            if (customer != null) {
                if (customer.customerId > 4L) {
                    println("*** Deleting customer = ${customer}")
                    repository.delete(customer)
                }
            }
        }
    }

    @Test
    fun testFindById() {
        val entity = service.getCustomerById(id)
        Assert.assertTrue(entity.isPresent)
        val (customerId, firstName, lastName, phone, email) = entity.get()
        Assert.assertEquals(id, customerId)
        Assert.assertEquals("Almost", firstName)
        Assert.assertEquals("There", lastName)
        Assert.assertEquals("", phone)
        Assert.assertEquals("", email)
    }

    @Test
    fun testFindAll() {
        val customers: List<CustomerDigest> = service.getCustomers()
        Assert.assertEquals(4, customers.size.toLong())
        for (customer in customers) {
            println("customer = $customer")
        }
    }

    @Test
    fun testUpdateCustomer() {
        val customer = CustomerDigest(
                customerId = id,
                firstName = "Yogi",
                lastName = "Bear",
                phone = "8005550001",
                email = "main.bear@jellystone.gov")

        val serviceResult = service.updateCustomer(customer)
        Assert.assertEquals(customer, serviceResult)
        val entity = repository.findById(id)
        Assert.assertTrue(entity.isPresent)
        assertEquals(customer, entity.get())
    }

    private fun assertEquals(customerDigest: CustomerDigest, customer: Customer) {
        Assert.assertEquals(customerDigest.customerId, customer.customerId)
        Assert.assertEquals(customerDigest.firstName, customer.firstName)
        Assert.assertEquals(customerDigest.lastName, customer.lastName)
        Assert.assertEquals(customerDigest.phone, customer.phone)
        Assert.assertEquals(customerDigest.email, customer.email)
    }

    @Test
    fun testUpdateCustomer_invalidId() {
        val invalidId: Long = 9999
        val customer = CustomerDigest(
                customerId = invalidId,
                firstName = "Yogi",
                lastName = "Bear",
                phone = "8005550001",
                email = "main.bear@jellystone.gov")

        val serviceResult = service.updateCustomer(customer)
        Assert.assertNull(serviceResult)
        val entity = repository.findById(invalidId)
        Assert.assertFalse(entity.isPresent)
    }

    @Test
    fun testAddAndDeleteCustomer() {
        val customer = CustomerDigest(firstName = "Teen", lastName = "Titans", phone = "8005557777",
                customerId = 999, email = "")
        val expectedCount = repository.count() + 1
        val customerDigest = service.addCustomer(customer)

        Assert.assertEquals(expectedCount, repository.count())

        customer.customerId = customerDigest.customerId
        Assert.assertEquals(customer, customerDigest)
        service.deleteCustomer(customerDigest.customerId)
        Assert.assertEquals(expectedCount - 1, repository.count())
        Assert.assertFalse(service.getCustomerById(customerDigest.customerId).isPresent)
    }
}
